tpu-client-next/src/node_address_service/leader_tpu_cache_service.rs

```diff
@@ -0,0 +1,520 @@
+
+                    if let Err(e) = leaders_sender.send((leaders, config.lookahead_leaders != lookahead_leaders)) {
+                        warn!("Unexpectedly dropped leaders_sender: {e}");
+                        return Err(Error::ChannelClosed);
+                    }
+                }
+
+                _ = cancel.cancelled() => {
+                    info!("Cancel signal received, stopping LeaderTpuCacheService.");
+                    break;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error(transparent)]
+    RpcError(#[from] ClientError),
+
+    #[error("Failed to get slot leaders connecting to: {0}")]
+    SlotLeadersConnectionFailed(String),
+
+    #[error("Failed find any cluster node info for upcoming leaders, timeout: {0}")]
+    ClusterNodeNotFound(String),
+
+    #[error(transparent)]
+    JoinError(#[from] tokio::task::JoinError),
+
+    #[error("Unexpectly dropped a channel.")]
+    ChannelClosed,
+
+    #[error("Failed to initialize LeaderTpuCacheService.")]
+    InitializationFailed,
+}
+
+#[async_trait]
+pub trait ClusterInfoProvider: Send + Sync {
+    async fn leader_tpu_map(&self) -> Result<HashMap<Pubkey, SocketAddr>, Error>;
+    async fn epoch_info(&self, estimated_current_slot: Slot) -> Result<(Slot, Slot), Error>;
+    async fn slot_leaders(
+        &self,
+        estimated_current_slot: Slot,
+        slots_in_epoch: Slot,
+    ) -> Result<Vec<Pubkey>, Error>;
+}
+
+async fn update_leader_info(
+    estimated_current_slot: Slot,
+    rpc_client: &impl ClusterInfoProvider,
+    epoch_info: &mut EpochInfo,
+    slot_leaders: &mut SlotLeaders,
+    num_consecutive_failures: &mut usize,
+    max_consecutive_failures: usize,
+) -> Result<(), Error> {
+    if estimated_current_slot > epoch_info.last_slot_in_epoch {
+        try_update(
+            "epoch info",
+            epoch_info,
+            || EpochInfo::new(rpc_client, estimated_current_slot),
+            num_consecutive_failures,
+            max_consecutive_failures,
+        )
+        .await?;
+    }
+    if estimated_current_slot.saturating_add(MAX_FANOUT_SLOTS) > slot_leaders.last_slot() {
+        try_update(
+            "slot leaders",
+            slot_leaders,
+            || {
+                SlotLeaders::new(
+                    rpc_client,
+                    estimated_current_slot,
+                    epoch_info.slots_in_epoch,
+                )
+            },
+            num_consecutive_failures,
+            max_consecutive_failures,
+        )
+        .await?;
+    }
+    Ok(())
+}
+
+/// Get the TPU sockets for the current and upcoming leaders according to
+/// fanout size.
+fn leader_sockets(
+    estimated_current_slot: Slot,
+    lookahead_leaders: u8,
+    slot_leaders: &SlotLeaders,
+    leader_tpu_map: &LeaderTpuMap,
+) -> Vec<SocketAddr> {
```

Comment:
I find it handy to pass around `(Pubkey, SocketAddr)` instead of only the addr

IK that's **significant**